### PR TITLE
PCHR-2235: Allow Public Holidays/Public Holiday Leave Requests With Past Dates To be Created

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
Currently creating a Public Holiday with past dates is not allowed as it would involve creating Public Holiday Leave requests for employees for that date (If MTPHL is true) and this will impact on things like Entitlements, Brought Forward, Expired balance Changes and so on, However there are some valid reasons why this may be necessary, one of such is when migrating data from an old system to use the new L&A it may be necessary to create past public holidays for contacts, also when creating a contract for an employee and the contract is created after the dates for some public holidays have passed but those public holidays are supposed to be included in the contract.

## Before
The system does not allow creating Public Holidays in the past and also does not allow creating past public holidays within a contract period when creating a Contract for a contact.

## After
The system was  modified to allow creating Public Holidays in the past and also to allow creating past public holiday leave requests within a contract period when creating a Contract for a contact. After a Public Holiday Leave request with past dates is created, recalculation of Expired Balance Changes is done to keep the calculation in sync.

## Comments
Creating a Public Holiday Leave Requests will require Entitlement Recalculation to be done as this has an impact on Contacts Entitlement, Nothing is done in the PR about this since the Entitlement Calculation is to be done manually by the user who made the change.
- [X] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChangeTest.php [hunk 1]
```diff
@@ -2408,6 +2408,11 @@ public function testRecalculateExpiredBalanceChangesWhenAbsenceTypeOfLeaveReques
   public function testCalculateAmountForDateForAContactWithWorkPatternHavingMultipleWeeks() {
     $periodStartDate = new DateTime('2016-01-01');
 
+    AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date' => CRM_Utils_Date::processDate('2016-12-31')
+    ]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChangeTest.php]: Why did you need to add the absence period on these tests? Is it because now the creation of Public Holidays will trigger the recalculation of expired records? If that is the case, what was the performance impact of this during the tests?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php [hunk 2]
```diff
@@ -531,4 +540,184 @@ public function testCreateLeaveRequestsForAllPublicHolidaysInTheFutureForDefault
...
+  public function testExpiredBalanceChangeIsRecalculatedForAllContactsWhenCreatingPublicHolidayWithPastDates() {
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-06-01'),
+      'end_date' => CRM_Utils_Date::processDate('+10 days')
+    ]);
+
+    $contact1 = ContactFabricator::fabricate();
+    $contact2 = ContactFabricator::fabricate();
+
+
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php]: please remove one of those 2 empty lines here

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php [hunk 3]
```diff
@@ -531,4 +540,184 @@ public function testCreateLeaveRequestsForAllPublicHolidaysInTheFutureForDefault
...
+      'period_id' => $absencePeriod->id,
+      'type_id' => $this->absenceType->id,
+    ]);
+
+    HRJobContractFabricator::fabricate([
+      'contact_id' => $contact1['id']
+    ],
+    [
+      'period_start_date' => CRM_Utils_Date::processDate('2016-06-01'),
+    ]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php]: can we please make this a bit more concise by using this formatting?
```php
HRJobContractFabricator::fabricate(
  [ 'contact_id' => $contact1['id'] ],
  [ 'period_start_date' => CRM_Utils_Date::processDate('2016-06-01') ]
);
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php [hunk 4]
```diff
@@ -531,4 +540,184 @@ public function testCreateLeaveRequestsForAllPublicHolidaysInTheFutureForDefault
...
+
+    $expiredBalanceChange1 = $this->getBalanceChangeRecord($balanceChange1->id);
+    //The public holiday date (2016-11-03) is before the date the balance change expired so one will be deducted
+    //from $balanceChange1
+    //Public Holiday2 is in past but the date is not before an already expired balance change
+    //Public Holiday3 is in the future, so it does not affect the recalculation
+    $this->assertEquals(-2, $expiredBalanceChange1->amount);
+  }
+
+  private function getBalanceChangeRecord($balanceChangeID) {
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php]: Couldn't you have just used `LeaveBalanceChange::findById($id)` instead?

Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php]: Why didn't I think of that

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChangeTest.php (hunk 1 above)
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChangeTest.php]: Yes absence period needs to be present when creating public holiday with past dates. See here [Here](https://github.com/civicrm/civihr/blob/c3fb4d1dca6a337b6f54d2337299792e6164c544/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php#L877-L877). There did not to be much performance difference as I ran tests both on my branch and also on the la-milestone-3 branch.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php (hunk 4 above)
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php]: well... my idea was to actually replace `getBalanceChangeRecord()` with `findById()`. Since now one method is just delegating the job to the other, there's really no reason for `getBalanceChangeRecord()` to exist
